pyccel/parser/semantic.py

```diff
@@ -234,8 +239,27 @@ def annotate(self, **settings):
 
         self._semantic_done = True
 
+        # check if the ast CodeBlock is a Program to call the Garbage collecting
+        cls = (Header, EmptyNode, NewLine, Comment, CommentBlock, Module)
+        is_module = all(isinstance(i,cls) for i in ast.body)
+        if not is_module:
+            self._ast = ast = self.garbage_collector(ast)
+        self._allocs = []
+
         return ast
 
+    def garbage_collector(self, expr):
+        """
+        Search in a CodeBlock if no trailing Return Node is present add the needed frees.
+
+        Return the same CodeBlock if a trailing Return found otherwise Return a new CodeBlock with additionals Deallocate Nodes.
+        """
+        if not isinstance(expr.body[-1], Return):
+            code = expr.body + [Deallocate(i) for i in self._allocs[-1]]
+            print(code)
+            return CodeBlock(code)
```

Comment:
Can't `self._allocs` be modified in this function?

Reply:
i will change the way i emove the last element and make it part of the function